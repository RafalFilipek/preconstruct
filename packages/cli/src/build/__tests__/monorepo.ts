import path from "path";
import fixturez from "fixturez";
import build from "../";
import {
  snapshotDirectory,
  testdir,
  repoNodeModules,
  initBasic,
  install,
  snapshotDistFiles,
} from "../../../test-utils";

jest.setTimeout(6000000);

jest.mock("../../prompt");

const f = fixturez(__dirname);

let unsafeRequire = require;
test("monorepo", async () => {
  let tmpPath = f.copy("monorepo");
  await initBasic(tmpPath);
  await install(tmpPath);
  await build(tmpPath);
  let counter = 1;
  for (let pkg of ["package-one", "package-two"]) {
    let pkgPath = path.join(tmpPath, "packages", pkg);
    await snapshotDistFiles(pkgPath);

    expect(unsafeRequire(pkgPath).default).toBe(counter++);
  }
});

test("monorepo with local .babelrc", async () => {
  let tmpPath = await testdir({
    "package.json": JSON.stringify({
      name: "monorepo-babelrc",
      preconstruct: {
        packages: ["packages/*"],
      },
    }),
    "packages/pkg/package.json": JSON.stringify({
      name: "monorepo-babelrc-pkg",
      main: "dist/monorepo-babelrc-pkg.cjs.js",
      module: "dist/monorepo-babelrc-pkg.esm.js",
    }),
    "packages/pkg/src/index.js": "export default () => <div>hello</div>",
    "packages/pkg/.babelrc": JSON.stringify({
      presets: [require.resolve("@babel/preset-react")],
    }),
    "packages/pkg/node_modules": {
      kind: "symlink",
      path: repoNodeModules,
    },
  });

  await build(tmpPath);

  await snapshotDirectory(path.join(tmpPath, "dist"), { files: "all" });
});

test("monorepo with two packages and one .babelrc", async () => {
  let tmpPath = await testdir({
    "package.json": JSON.stringify({
      name: "monorepo-babelrc",
      preconstruct: {
        packages: ["packages/*"],
      },
    }),
    "packages/pkg/package.json": JSON.stringify({
      name: "monorepo-babelrc-pkg",
      main: "dist/monorepo-babelrc-pkg.cjs.js",
      module: "dist/monorepo-babelrc-pkg.esm.js",
    }),
    "packages/pkg/src/index.js": "export default () => <div>hello</div>",
    "packages/pkg/.babelrc": JSON.stringify({
      presets: [require.resolve("@babel/preset-react")],
    }),
    "packages/pkg/node_modules": {
      kind: "symlink",
      path: repoNodeModules,
    },
    "packages/pkg2/package.json": JSON.stringify({
      name: "monorepo-babelrc-pkg2",
      main: "dist/monorepo-babelrc-pkg2.cjs.js",
      module: "dist/monorepo-babelrc-pkg2.esm.js",
      dependencies: {
        "monorepo-babelrc-pkg": "*",
      },
    }),
    "packages/pkg2/src/index.js": `
      import Component from "monorepo-babelrc-pkg";

      export default Component;
    `,
  });

  await build(tmpPath);

  await snapshotDirectory(path.join(tmpPath, "dist"), { files: "all" });
});

test("monorepo with missing .babelrc", async () => {
  let tmpPath = await testdir({
    "package.json": JSON.stringify({
      name: "monorepo-babelrc",
      preconstruct: {
        packages: ["packages/*"],
      },
    }),
    "packages/pkg/package.json": JSON.stringify({
      name: "monorepo-babelrc-pkg",
      main: "dist/monorepo-babelrc-pkg.cjs.js",
      module: "dist/monorepo-babelrc-pkg.esm.js",
    }),
    "packages/pkg/src/index.js": "export default () => <div>hello</div>",
    "packages/pkg/.babelrc": JSON.stringify({
      presets: [require.resolve("@babel/preset-react")],
    }),
    "packages/pkg/node_modules": {
      kind: "symlink",
      path: repoNodeModules,
    },
    "packages/pkg2/package.json": JSON.stringify({
      name: "monorepo-babelrc-pkg2",
      main: "dist/monorepo-babelrc-pkg2.cjs.js",
      module: "dist/monorepo-babelrc-pkg2.esm.js",
      dependencies: {
        "monorepo-babelrc-pkg": "*",
      },
    }),
    "packages/pkg2/src/index.js": `
      export default () => <div>hello</div>;
    `,
  });

  await expect(build(tmpPath)).rejects.toThrow();
});
