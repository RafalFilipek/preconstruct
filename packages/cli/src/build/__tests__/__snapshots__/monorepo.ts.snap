// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`monorepo: some-scope-package-one.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index = 1;

exports.default = index;
"
`;

exports[`monorepo: some-scope-package-one.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === \\"production\\") {
  module.exports = require(\\"./some-scope-package-one.cjs.prod.js\\");
} else {
  module.exports = require(\\"./some-scope-package-one.cjs.dev.js\\");
}
"
`;

exports[`monorepo: some-scope-package-one.cjs.prod.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index = 1;

exports.default = index;
"
`;

exports[`monorepo: some-scope-package-one.esm.js 1`] = `
"var index = 1;

export default index;
"
`;

exports[`monorepo: some-scope-package-two.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index = 2;

exports.default = index;
"
`;

exports[`monorepo: some-scope-package-two.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === \\"production\\") {
  module.exports = require(\\"./some-scope-package-two.cjs.prod.js\\");
} else {
  module.exports = require(\\"./some-scope-package-two.cjs.dev.js\\");
}
"
`;

exports[`monorepo: some-scope-package-two.cjs.prod.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index = 2;

exports.default = index;
"
`;

exports[`monorepo: some-scope-package-two.esm.js 1`] = `
"var index = 2;

export default index;
"
`;
